File: gtci/12-tree-depth-first-search/python3/8-path-with-maximum-sum.py
import math


class TreeNode:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    def findMaximumPathSum(self, root):
        self.globalMaximumSum = -math.inf

        self.findMaximumPathSum_helper(root)

        return self.globalMaximumSum

    def findMaximumPathSum_helper(self, node):
        if node is None:
            return 0

        # max of 0 and a positive path
        # we would not include the negative branch in our calculation of the global max sum
        max_sum_left = max(0, self.findMaximumPathSum_helper(node.left))
        max_sum_right = max(0, self.findMaximumPathSum_helper(node.right))

        self.globalMaximumSum = max(
            self.globalMaximumSum, max_sum_left + max_sum_right + node.val
        )

        return max(max_sum_left, max_sum_right) + node.val


if __name__ == "__main__":
    sol = Solution()

    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    assert sol.findMaximumPathSum(root) == 6

    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.right.left = TreeNode(5)
    root.right.right = TreeNode(6)
    assert sol.findMaximumPathSum(root) == 16

    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(1)
    root.left.right = TreeNode(3)
    root.right.left = TreeNode(5)
    root.right.right = TreeNode(6)
    root.right.left.left = TreeNode(7)
    root.right.left.right = TreeNode(8)
    root.right.right.right = TreeNode(9)
    assert sol.findMaximumPathSum(root) == 31

    root = TreeNode(-1)
    root.left = TreeNode(-3)
    assert sol.findMaximumPathSum(root) == -1

    root = TreeNode(10)
    root.left = TreeNode(-5)
    root.right = TreeNode(3)
    assert sol.findMaximumPathSum(root) == 13

    print("All test cases passed.")

File: gtci/12-tree-depth-first-search/cpp/8-path-with-maximum-sum.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <limits>

#include "tree_utils.h"

using namespace std;

class Solution {
public:
  int findMaximumPathSum(TreeNode* root) {
    if (root == nullptr)
      return 0;

    int globalMaximumSum = numeric_limits<int>::min();
    
    findMaximumPathSumHelper(root, globalMaximumSum);

    return globalMaximumSum;
  }

  int findMaximumPathSumHelper(TreeNode* node, int& globalMaximumSum) {
    if (node == nullptr)
      return 0;
    
    const int leftMaximumSum = std::max(0, findMaximumPathSumHelper(node->left, globalMaximumSum));
    const int rightMaximumSum = std::max(0, findMaximumPathSumHelper(node->right, globalMaximumSum));

    globalMaximumSum = std::max(globalMaximumSum, leftMaximumSum + rightMaximumSum + node->val);

    return std::max(leftMaximumSum, rightMaximumSum) + node->val;
  }
};

int main() {
    Solution sol;

    TreeNode* root1 = new TreeNode(1);
    root1->left = new TreeNode(2);
    root1->right = new TreeNode(3);
    assert(sol.findMaximumPathSum(root1) == 6);

    TreeNode* root2 = new TreeNode(1);
    root2->left = new TreeNode(2);
    root2->right = new TreeNode(3);
    root2->left->left = new TreeNode(4);
    root2->right->left = new TreeNode(5);
    root2->right->right = new TreeNode(6);
    assert(sol.findMaximumPathSum(root2) == 16);

    TreeNode* root3 = new TreeNode(1);
    root3->left = new TreeNode(2);
    root3->right = new TreeNode(3);
    root3->left->left = new TreeNode(1);
    root3->left->right = new TreeNode(3);
    root3->right->left = new TreeNode(5);
    root3->right->right = new TreeNode(6);
    root3->right->left->left = new TreeNode(7);
    root3->right->left->right = new TreeNode(8);
    root3->right->right->right = new TreeNode(9);
    assert(sol.findMaximumPathSum(root3) == 31);

    TreeNode* root4 = new TreeNode(-1);
    root4->left = new TreeNode(-3);
    assert(sol.findMaximumPathSum(root4) == -1);

    TreeNode* root5 = new TreeNode(10);
    root5->left = new TreeNode(-5);
    root5->right = new TreeNode(3);
    assert(sol.findMaximumPathSum(root5) == 13);

    cout << "All test cases passed." << endl;
    
    delete root1;
    delete root2;
    delete root3;
    delete root4;
    delete root5;
    
    return 0;
}
